feat(client): sign-up #5

shinjigu opened the pull request:
sign up func

- Form Input : 이름, 아이디, 닉네임, 비밀번호, 비밀번호 확인 
- sign-up button
- sign-in link

실제 서버와의 연결 X - 아이디/닉네임/이메일 중복체크는 현재 모두 true 반환

기타 변경 사항
- form-input.tsx
validMessage prop 추가해서 입력값이 유효할 때 성공 메시지 표시
유효성 검사 통과 시 accent 색상, 실패 시 destructive 색상 적용
- form.tsx
Label 스타일링 변경 - error 상태일 때 텍스트 색상 변경되던 기능 제거

shinjigu committed:
feat(client): sign-up #5

sign up func

Form Input
- 이름, 아이디, 닉네임, 비밀번호, 비밀번호 확인
sign-up button
sign-in link

실제 서버와의 연결 X - 아이디/닉네임/이메일 중복체크는 현재 모두 true 반환

기타 변경 사항
- form-input.tsx
validMessage prop 추가해서 입력값이 유효할 때 성공 메시지 표시
유효성 검사 통과 시 accent 색상, 실패 시 destructive 색상 적용

- form.tsx
Label 스타일링 변경 - error 상태일 때 텍스트 색상 변경되던 기능 제거

DonghyunKim98 commented on apps/web/src/app/sign-up/layout.tsx:
```diff
@@ -0,0 +1,15 @@
+import { ReactNode } from 'react';
+import Image from 'next/image';
+
+export default function Layout({ children }: { children: ReactNode }) {
```
- Props 의 경우 따로 type 을 지정해주는 편이 유지보수성과 확장성면에서 좋습니다.
- React 에는 `ChildrenWithProps` 라는 타입이 존재합니다.

```suggestion
type Props = PropsWithChildren<{}>

export default function Layout({ children }: Props) {
```

DonghyunKim98 commented on apps/web/src/components/common/form-input.tsx:
```diff
@@ -8,26 +8,32 @@ import {
   className,
   ...props
 }: FormInputProps<T>) {
+  const isFieldValid = form.getFieldState(name).isDirty && !form.getFieldState(name).error;
```
- `react-hook-form` 을 사용한다면 `useFormContext` 를 이용하여 form 을 굳이 props 로 받지 않아도 됩니다. (권장되는 스타일입니다. [관련 링크](https://react-hook-form.com/docs/useformcontext)) 
-  또한 Valid 한지를 확인하는 것 역시 useForm 에서 제공하는 `formState` 안에 isDirty 와 isValid 가 모두 있으니 이를 활용하시는 편이 좋습니다.

shinjigu replied:
- 첫 번째 리뷰 사항에서는 props에서 form을 제거하고, 컴포넌트에서 form prop 대신 useFormContext 사용하면 될까요 ?
- 그리고 두 번째에서는 getFieldState()를 formState 속성으로 대체하면 될까요 ?

DonghyunKim98 commented on apps/web/src/components/common/form-input.tsx:
```diff
@@ -39,7 +45,12 @@ export function FormInput<T extends FieldValues>({
+              !isFieldValid && "text-destructive" // 실패시 destructive 컬러
+            )}>
+              {isFieldValid ? validMessage : undefined}
+            </FormMessage>
```
```suggestion
           {isFieldValid &&(
              <FormMessage .../>
           )}
```

`isFieldValid` 할때 `FormMessage` 를 렌더링 하는 방식으로 가는건 어떨까요?

DonghyunKim98 commented on apps/web/src/components/containers/sign-up/sign-up.container.tsx:
```diff
@@ -0,0 +1,120 @@
+import { zodResolver } from '@hookform/resolvers/zod';
+import SignUpPresenter from './sign-up.presenter';
+
+const formSchema = z.object({
```
schema 는 따로 파일을 분리하는 편이 파일 가독성과 응집성 측면에서 낫습니다.
굳이 Container 가 관련 로직을 일일히 읽을 필요는 없어보입니다.

DonghyunKim98 commented on apps/web/src/components/containers/sign-up/sign-up.container.tsx:
```diff
@@ -0,0 +1,120 @@
+  id: z
+    .string()
+    .min(4, { message: '아이디는 4-12자 사이로 입력 가능합니다.' })
+    .max(12, { message: '아이디는 4-12자 사이로 입력 가능합니다.' })
```
1,6 같은 숫자들 constant 처리 해주시고, message 도 constant 처리해주세요.

DonghyunKim98 replied:
아래쪽 regex 도 마찬가지입니다.

DonghyunKim98 commented on apps/web/src/components/containers/sign-up/sign-up.container.tsx:
```diff
@@ -0,0 +1,120 @@
+    mode: 'onChange'
+  });
+
+  const formState = form.formState;
```
이렇게 쓰지말고 deconstruct 로 쓰는건 어떨까요?

```suggestion
  const {formState} = form;
```

DonghyunKim98 commented on apps/web/src/components/containers/sign-up/sign-up.container.tsx:
```diff
@@ -0,0 +1,120 @@
+    // 에러 없는지 확인
+    Object.keys(formState.errors).length === 0 &&
+    formState.isValid
+  );
```
- 런타임 유효성 검사에서 required 가 들어가 있다면 dirtyFields 를 검사하지 않아도 됩니다.

DonghyunKim98 replied:
`z.string().required()` 등으로 바꾸어 확인해보세요

DonghyunKim98 commented on apps/web/src/components/containers/sign-up/sign-up.container.tsx:
```diff
@@ -0,0 +1,120 @@
+  const onSubmit = () => {
+    console.log(form.getValues());
+    router.push('/sign-in');
+  };
```
onSubmit 의 경우 첫번째 파라미터로 입력된 데이터를 받을 수 있습니다. [관련 링크](https://react-hook-form.com/docs/useform/handlesubmit)
TypeScript 를 잘사용한 경우 자동으로 type 이 잡힙니다.

DonghyunKim98 commented on apps/web/src/components/containers/sign-up/sign-up.presenter.tsx:
```diff
@@ -0,0 +1,109 @@
+  isValid: boolean;
+}
+
+export default function SignUpPresenter({ form, onSubmit, isValid }: SignUpPresenterProps) {
```
- useController 나 useFormContext 를 이용해 props drilling 을 줄이는 편이 나은것 같습니다.
- useController 를 사용한다고 Container-Presenter 의 패턴을 해치는건 아니라고 생각합니다. 직접적인 로직은 Container 에서 선언이 되어있고 실제로는 추상화된 메소드만 가져다 쓰는 것이기 때문입니다.

DonghyunKim98 reviewed (changes requested):
수고하셨습니다. 처음 리뷰인데, 어려운 부분이 있으시면 알려주세요. 시간 나는대로 더 리뷰 해드리겠습니다.
react-hook-form 사용법이 아직 익숙치 않으신것 같습니다. `useFormContext` 를 이용하여 선언적+추상적으로 props 를 내려주지 않아도 사용이 가능하니 잘 적용해보세요

shinjigu committed:
feat(client): sign-up #5

동현님 리뷰 사항 반영하여 수정했습니다 !

DonghyunKim98 commented on apps/web/src/components/containers/sign-up/sign-up.container.tsx:
```diff
@@ -0,0 +1,44 @@
+
+  const {formState} = form;
+
+  const isValid = Boolean(
+    // 에러 없는지 확인
+    Object.keys(formState.errors).length === 0 &&
+    formState.isValid
+  );
```
fromState.isValid 만으로 확인할 수 있을 것 같아요!

DonghyunKim98 commented on apps/web/src/components/containers/sign-up/sign-up.container.tsx:
```diff
@@ -0,0 +1,44 @@
+    formState.isValid
+  );
+
+  const onSubmit = form.handleSubmit((data) => {
+    console.log(data);
+    router.push('/sign-in');
+  });
```
form.handleSubmit 보단 `handleSubmit` 으로 나타내도 될것 같습니다.

DonghyunKim98 replied:
react-hook-form 의 경우 전반적으로 deconstruct 를 해서 쓰는 편을 선호해요

standardwish replied:
동의합니다

standardwish commented on apps/web/src/components/containers/sign-up/schema.ts:
```diff
@@ -0,0 +1,103 @@
+  }
+ } as const;
+ 
+ export const formSchema = z.object({
```
혜진누나와 마찬가지로 변수 이름에 명확하게 `signUpFormSchema`와 같이 명시해주는 게 좋을 거 같습니다! 다른 `formSchema`와 혼용될 여지가 있어보여요

standardwish commented on apps/web/src/components/containers/sign-up/sign-up.presenter.tsx:
```diff
@@ -0,0 +1,96 @@
+              validMessage="사용 가능한 이메일입니다."
+            />
+
+            <Button type='submit' disabled={!isValid} className="w-full ${!isValid ? 'opacity-50 cursor-not-allowed' : ''}">
```
요런 식으로 className 적용 안되지 않나요???
```javascript
className={`w-full ${!isValid ? 'opacity-50 cursor-not-allowed' : ''}`}
```
위와 같이 써야 적용될텐디

그리고 button component에서 기본적으로 disabled 되었을 때 적용되는 효과가 있어 만약 피그마와 효과가 다르다면 컴포넌트 딴에서 수정해야 할 것 같아요~

standardwish commented on apps/web/src/components/containers/sign-up/sign-up.container.tsx:
아직 닉네임 검증함수가 안들어갔네요? 요거는 서버하고 맞춘 후에 다시 만들어야 할 거 같긴 한데, 그래도 어느정도 미리 고려하고 코드 작성해야 할 거 같습니다!

jisusin replied:
`export const validateId = async (id: string) => {
  try {
    // 서버 연동 후 실제 API 호출
    return true; // 임시 리턴
  } catch (error) {
    return false;
  }
};`
이런 느낌으로 틀만 잡아 놓고 나중에 연결하면 될까요 ?

standardwish reviewed (changes requested):
닉네임 검증 함수는 제가 서버 팀에 전달하고 의견 물어보겠습니다! 그 전까지 나머지 항목 수정 부탁드려요~

shinjigu committed:
feat(client): sign-up #5

동현님, 준성님 리뷰 사항 반영하여 수정했습니다 !

DonghyunKim98 commented on apps/web/src/components/common/form-input.tsx:
```diff
@@ -10,24 +11,29 @@ import {
   className,
   ...props
 }: FormInputProps<T>) {
+  const form = useFormContext<T>();
+  const { dirtyFields, errors } = form.formState;
+  const isFieldValid = dirtyFields[name as keyof typeof dirtyFields] && !errors[name];
```
- 잘 동작하나요? 상위에 `FormProvider` 를 선언 안 하신것 같아 여쭤봅니다.
```suggestion
  const {formState} = useFormContext<T>();
  const { dirtyFields, errors } = formState;
  const isFieldValid = !errors[name];
```
- 이렇게만 써도 충분해요. `dirtyFields` 의 경우에는 required option 으로 error 에서 체크합니다.

DonghyunKim98 commented on apps/web/src/components/common/form-input.tsx:
```diff
@@ -39,7 +45,14 @@ export function FormInput<T extends FieldValues>({
           </FormControl>
           <div className='min-h-[20px]'>
-            <FormMessage />
+            {isFieldValid && (
+              <FormMessage className="text-accent">
+                {validMessage}
+                </FormMessage>
+            )}
+            {!isFieldValid && (
+              <FormMessage className="text-destructive" />
+            )}
```
- 삼항연산자로 처리하는게 더 깔끔할것 같아요.

DonghyunKim98 commented on apps/web/src/components/containers/sign-up/schema.ts:
```diff
@@ -0,0 +1,130 @@
+  }
+ } as const;
+ 
+ export const signUpformSchema = z.object({
```
schema 자체에 required option 을 넣어서 dirtyFields 로 검사하는 이중 로직을 세우지 맙시다.

DonghyunKim98 replied:
`z.string().required()` 
이런식으로 선언해요.

DonghyunKim98 commented on apps/web/src/components/containers/sign-up/sign-up.container.tsx:
```diff
@@ -0,0 +1,40 @@
+  });
+
+  return (
+    <FormProvider {...form}>
```
여기 있었군요!

standardwish commented on apps/web/src/components/containers/sign-up/sign-up.container.tsx:
```diff
@@ -0,0 +1,40 @@
+  const isValid = Boolean(formState.isValid);
+
+  const onSubmit = handleSubmit((data) => {
+    console.log(data);
```
마지 전에 지워주세요~

standardwish reviewed (approved):
LGTM

shinjigu committed:
feat(client): sign-up #5

폼 입력 필드의 validation 메시지 표시 로직 수정
- watch 훅으로 현재 필드 값 확인
- 값이 있는 경우에만 validation 메시지 표시

DonghyunKim98 reviewed (approved):
수고하셨습니다 👍
